### ISSUE TITLE
Can't stream() from redirected file via stdin

### ISSUE THREAD

**Issue by katjonathan** [Gallopsled/pwntools#2036]
When trying to read from redirected stdin (as in the simple program below), Python throws an error: `AttributeError: 'NoneType' object has no attribute 'closed'`

```
#!/usr/bin/python3

from pwn import *

infile = 'tmpin.txt'
with open(infile, 'w') as fd:
	fd.write('bleh\n')

try:
	fd = os.open(infile, os.O_RDWR)
	io = process(['cat'], stdin=fd)
	io.stream()
finally:
	io.close()
	os.close(fd)
```
How can I successfully redirect the file "tmpin.txt" to the `cat` process?

**Comment by Arusekk** [comment]
What is the stacktrace that Python shows? In most cases the stack trace is more relevant than the message itself (at least in Python).

**Comment by katjonathan** [comment]
```
[+] Starting local process '/usr/bin/cat': pid 259
[*] Process '/usr/bin/cat' stopped with exit code 0 (pid 259)
Traceback (most recent call last):
  File "./play.py", line 12, in <module>
    io.stream()
  File "/usr/local/lib/python3.8/dist-packages/pwnlib/tubes/tube.py", line 920, in stream
    buf.add(function())
  File "/usr/local/lib/python3.8/dist-packages/pwnlib/tubes/tube.py", line 490, in recvline
    return self.recvuntil(self.newline, drop = not keepends, timeout = timeout)
  File "/usr/local/lib/python3.8/dist-packages/pwnlib/tubes/tube.py", line 333, in recvuntil
    res = self.recv(timeout=self.timeout)
  File "/usr/local/lib/python3.8/dist-packages/pwnlib/tubes/tube.py", line 105, in recv
    return self._recv(numb, timeout) or b''
  File "/usr/local/lib/python3.8/dist-packages/pwnlib/tubes/tube.py", line 183, in _recv
    if not self.buffer and not self._fillbuffer(timeout):
  File "/usr/local/lib/python3.8/dist-packages/pwnlib/tubes/tube.py", line 154, in _fillbuffer
    data = self.recv_raw(self.buffer.get_fill_size())
  File "/usr/local/lib/python3.8/dist-packages/pwnlib/tubes/process.py", line 685, in recv_raw
    self.shutdown("recv")
  File "/usr/local/lib/python3.8/dist-packages/pwnlib/tubes/tube.py", line 1236, in shutdown
    self.shutdown_raw(self.shutdown_directions[direction])
  File "/usr/local/lib/python3.8/dist-packages/pwnlib/tubes/process.py", line 771, in shutdown_raw
    if False not in [self.proc.stdin.closed, self.proc.stdout.closed]:
AttributeError: 'NoneType' object has no attribute 'closed'
```

**Comment by goreil** [comment]
To narrow down the issue, the issue seems to be with the method `stream()` combined with redirecting `stdin`:

We can see that the code
``` python3
from pwn import *
p = process(['/usr/bin/cat', 'test.txt'])
p.stream()
```
works perfectly fine, but

```python3
from pwn import *
fd = open('test.txt', 'r')
p = process(['/usr/bin/cat'], stdin = fd)
p.stream()
```

Throws the mentioned error.

**Comment by katjonathan** [comment]
Well, yes. But in truth, I actually needed to run a custom CTF binary that requires a file to be redirected to stdin. It's just that a custom binary would make for a poor choice for a GitHub question. I found that replacing the binary with any other binary (such as `cat` or `echo`) also throws the same error, so that works better as a minimum viable product to demonstrate the error thrown. I've also found that changing `stream()` to `recvall()` (as in `print(p.recvall())`) also runs into the same error.

**Comment by katjonathan** [comment]
Should this be classified as a bug or am I merely misusing the library? If it is a misuse case, how should I properly redirect input from a file?

**Comment by caprinux** [comment]
This is indeed a bug, with some basic debugging I quickly arrived at this line in `process.py`.

```py
# pwnlib/tubes/process.py (_init.py)
    279         # Create the PTY if necessary
    280         stdin, stdout, stderr, master, slave = self._handles(*handles)
```

```py
# pwnlib/tubes/process.py (_handles)
    598             if stdin is PTY:
# Problem is here because stdin is a `TextIOWrapper` object and not a `PTY`
    599                 stdin = slave
    600             if stdout is PTY:
    601                 stdout = slave
    602             if stderr is PTY:
    603                 stderr = slave
```

```py
# pwnlib/tubes/process.py (_init.py)
    348         if self.pty is not None:
    349             if stdin is slave:
    350                 self.proc.stdin = os.fdopen(os.dup(master), 'r+b', 0)
    351             if stdout is slave:
    352                 self.proc.stdout = os.fdopen(os.dup(master), 'r+b', 0)
    353             if stderr is slave:
    354                 self.proc.stderr = os.fdopen(os.dup(master), 'r+b', 0)
```

There is a need to fix the underlying problem of defining `self.proc.stdin` despite the current logic trying to establish a PTY to provide the communication channel.

Could implement a simple fix to run `self.proc.stdin = os.fdopen(os.dup(master), 'r+b', 0)` after a check if `fd == TextIOWrapper`, but I would rather leave that for the developers to implement in the case that there are underlying problems that I may have missed. A fix could look something like this

```py
# pwnlib/tubes/process.py (_init.py)
                import io 
                ...
    348         if self.pty is not None:
    349             if stdin is slave:
    350                 self.proc.stdin = os.fdopen(os.dup(master), 'r+b', 0)
                    elif isinstance(stdin, io.TextIOWrapper):
                        self.proc.stdin = os.fdopen(os.dup(master), 'r+b', 0)
    351             if stdout is slave:
    352                 self.proc.stdout = os.fdopen(os.dup(master), 'r+b', 0)
    353             if stderr is slave:
    354                 self.proc.stderr = os.fdopen(os.dup(master), 'r+b', 0)
```

**Comment by caprinux** [comment]
I could submit a PR for this if any pwntools maintainer could have a look at this and drop a comment :)

**Comment by katjonathan** [comment]
That would be awesome!

**Comment by jamestiotio** [comment]
Hi @katjonathan, it seems that I have fixed this issue in #2102. While we wait for a maintainer to review said PR, you can attempt to use my fork and see whether it indeed fixes this issue on your side. I have tested the fix myself and it seems to have fixed this issue, at least on my side.

Cheers!
